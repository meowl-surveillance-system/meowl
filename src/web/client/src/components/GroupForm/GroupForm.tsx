import React, {
  Component,
  FormEvent,
  ChangeEvent,
  ComponentState,
} from "react";
import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import Container from "@material-ui/core/Container";

interface Props {}
interface State {
  groupId: string;
  username: string;
  successMessage: string;
  failMessage: string;
}

/**
 * A component for adding a user to a group
 */
export default class GroupForm extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      groupId: "",
      username: "",
      successMessage: "",
      failMessage: "",
    };
  }

  /**
   * Submits the inputted groupId and username to the backend to add the user to a group
   * If successful, a success message will be displayed, otherwise a fail message will be displayed.
   * @params event - An event that records submission of a form element
   */
  onSubmit = async (event: FormEvent<HTMLFormElement>) => {
    try {
      event.preventDefault();
      const requestOptions = {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          groupId: this.state.groupId,
          username: this.state.username
        }),
      };
      const res = await fetch("/api/addUserGroup", requestOptions);
      const msg = await res.text();
      if (msg === "Successfully added user to group") {
        this.setState({ successMessage: msg, failMessage: "" });
      } else {
        this.setState({ failMessage: msg, successMessage: "" });
      }
    } catch (e) {
      console.error(e);
    }
  };

  /**
   * Sets the appropriate state when a keyboard event is triggered
   * @params event - An event that records changes happening to an input field
   */
  onChange = async (event: ChangeEvent<HTMLInputElement>) => {
    event.preventDefault();
    this.setState({
      [event.target.name]: event.target.value,
    } as ComponentState);
  };

  /**
   * Renders out a form for the user to input the new password
   */
  render() {
    return (
      <Container component="main" maxWidth="xs">
          {this.state.failMessage && (
            <Typography style={{ color: "red" }}>
              {this.state.failMessage}
            </Typography>
          )}
          {this.state.successMessage && (
            <Typography style={{ color: "green" }}>
              {this.state.successMessage}
            </Typography>
          )}
          <Typography component="h1" variant="h5">
            Add User to Group
          </Typography>
          <Typography>
            {"Add user to a group. If groupId does not exist, a new group will be created with the provided groupId"}
          </Typography>
          <form noValidate onSubmit={this.onSubmit}>
            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              name="groupId"
              label="GroupId"
              onChange={this.onChange}
            />
            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              name="username"
              label="Username"
              onChange={this.onChange}
            />
            <Button fullWidth type="submit" variant="contained" color="primary">
              Submit
            </Button>
          </form>
      </Container>
    );
  }
}
