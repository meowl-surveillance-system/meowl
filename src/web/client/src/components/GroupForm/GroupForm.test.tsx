import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import GroupForm from "./GroupForm";

describe("GroupForm component", () => {
  let GroupFormWrapper: ShallowWrapper<{}, {}, GroupForm>;
  let eventMock: any;
  let preventDefaultMock: jest.Mock;
  let mockSuccessResponse: string;
  let mockTextPromise: Promise<string>;
  let mockFetchPromise: Promise<object>;
  beforeEach(() => {
    mockSuccessResponse = "Successfully added user to group";
    mockTextPromise = Promise.resolve(mockSuccessResponse);
    mockFetchPromise = Promise.resolve({ text: () => mockTextPromise });
    jest
      .spyOn(window, "fetch")
      .mockImplementation(() => mockFetchPromise as Promise<Response>);
    preventDefaultMock = jest.fn();
    eventMock = {
      preventDefault: preventDefaultMock,
      target: {
        name: "groupId",
        value: "muda",
      },
    };
    GroupFormWrapper = shallow(<GroupForm />);
  });
  test("renders GroupForm component", () => {
    const { container } = render(
      <Router>
        <GroupForm />
      </Router>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should set successMessage on successful resolve", async () => {
    await GroupFormWrapper.instance().onSubmit(eventMock);
    expect(GroupFormWrapper.instance().state.successMessage).toEqual(
      mockSuccessResponse,
    );
  });
  it("should not set failMessage on promise reject", async () => {
    await GroupFormWrapper.instance().onSubmit(eventMock);
    expect(GroupFormWrapper.instance().state.failMessage).toEqual("");
  });
  it("should set groupId field when onChange is called", () => {
    GroupFormWrapper.instance().onChange(eventMock);
    expect(GroupFormWrapper.instance().state.groupId).toEqual(
      eventMock.target.value,
    );
  });
});
