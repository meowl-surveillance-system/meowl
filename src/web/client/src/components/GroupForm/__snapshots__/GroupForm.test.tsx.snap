// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupForm component renders GroupForm component 1`] = `
<main
  class="MuiContainer-root MuiContainer-maxWidthXs"
>
  
  
  <h1
    class="MuiTypography-root MuiTypography-h5"
  >
    Add User to Group
  </h1>
  <p
    class="MuiTypography-root MuiTypography-body1"
  >
    Add user to a group. If groupId does not exist, a new group will be created with the provided groupId
  </p>
  <form
    novalidate=""
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
        data-shrink="false"
      >
        GroupId
        <span
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          name="groupId"
          required=""
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-97 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-99"
          >
            <span>
              GroupId
               *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
        data-shrink="false"
      >
        Username
        <span
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          name="username"
          required=""
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-97 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-99"
          >
            <span>
              Username
               *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label"
      >
        Submit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </form>
</main>
`;
